packages/grafana-data/src/transformations/transformers/joinDataFrames.ts

```diff
@@ -0,0 +1,293 @@
+        join = field;
+      } else {
+        if (options.keep && !options.keep(field, frame, options.frames)) {
+          continue; // skip field
+        }
+
+        // Support the standard graph span nulls field config
+        nullModesFrame.push(field.config.custom?.spanNulls ? NULL_REMOVE : NULL_EXPAND);
+
+        let labels = field.labels ?? {};
+        if (frame.name) {
+          labels = { ...labels, name: frame.name };
+        }
+
+        fields.push({
+          ...field,
+          labels, // add the name label from frame
+        });
+      }
+
+      if (options.keepOriginIndices) {
+        field.state!.origin = {
+          frameIndex,
+          fieldIndex,
+        };
+      }
+    }
+
+    if (!join) {
+      continue; // skip the frame
+    }
+
+    if (originalFields.length < 1) {
+      originalFields.push(join); // first join field
+    }
+    nullModes.push(nullModesFrame);
+
+    const a: AlignedData = [join.values.toArray()]; //
+    for (const field of fields) {
+      a.push(field.values.toArray());
+      originalFields.push(field);
+    }
+    allData.push(a);
+  }
+
+  const joined = join(allData, nullModes);
+  if (!joined) {
+    return undefined;
+  }
+
+  return {
+    // ...options.data[0], // keep name, meta?
+    length: joined[0].length,
+    fields: originalFields.map((f, index) => ({
+      ...f,
+      values: new ArrayVector(joined[index]),
+    })),
+  };
+}
+
+//--------------------------------------------------------------------------------
+// Below here is copied from uplot (MIT License)
+// https://github.com/leeoniya/uPlot/blob/master/src/utils.js#L325
+// This avoids needing to import uplot into the data package
+//--------------------------------------------------------------------------------
+
+// Copied from uplot
+type AlignedData = [number[], ...Array<Array<number | null>>];
+
+// nullModes
+const NULL_REMOVE = 0; // nulls are converted to undefined (e.g. for spanGaps: true)
+const NULL_RETAIN = 1; // nulls are retained, with alignment artifacts set to undefined (default)
+const NULL_EXPAND = 2; // nulls are expanded to include any adjacent alignment artifacts
+
+type JoinNullMode = number; // NULL_IGNORE | NULL_RETAIN | NULL_EXPAND;
+
+// sets undefined values to nulls when adjacent to existing nulls (minesweeper)
+function nullExpand(yVals: Array<number | null>, nullIdxs: number[], alignedLen: number) {
+  for (let i = 0, xi, lastNullIdx = -1; i < nullIdxs.length; i++) {
+    let nullIdx = nullIdxs[i];
+
+    if (nullIdx > lastNullIdx) {
+      xi = nullIdx - 1;
+      while (xi >= 0 && yVals[xi] == null) {
+        yVals[xi--] = null;
+      }
+
+      xi = nullIdx + 1;
+      while (xi < alignedLen && yVals[xi] == null) {
+        yVals[(lastNullIdx = xi++)] = null;
+      }
+    }
+  }
+}
+
+// nullModes is a tables-matched array indicating how to treat nulls in each series
+function join(tables: AlignedData[], nullModes: number[][]) {
+  if (tables.length === 1) {
```

Comment:
let's kill this early exit block to get uniform nullModes behavior for all inputs. the single frame bailout should happen outside of this fn (one level up), i think.